opte/src/engine/checksum.rs

```diff
@@ -59,10 +59,32 @@ impl Checksum {
             self.inner = (self.inner >> 16) + (self.inner & 0xFFFF);
         }
 
-        ((self.inner & 0xFFFF) as u16).to_ne_bytes()
+        let sum = (self.inner & 0xFFFF) as u16;
+        if sum == 0 {
+            [0xFF, 0xFF]
+        } else {
+            sum.to_ne_bytes()
+        }
     }
```

Comment:
This change is incorrect as it's happening at the wrong time. You'll never trigger this case, because a rolling checksum of all zeros is only possible if the header is all zero bytes.

The `Checksum` type is used to store a rolling one's complement checksum calculation, it **is not** the checksum value stored in the header. The `HeaderChecksum` type is the type that is actually stored/written to a header (I have a todo in my local branch to document these, but I should probably float an individual PR for it). If you look at its `From` impl you'll see it performs bitwise negation (one's complement) of the current sum, which is the final step. It's **that value** that needs the check against `0x0000`, which is a totally valid result for a TCP checksum, but results in ambiguity for a UDP checksum.

I think we should apply this special case only to the UDP code after it computes its `HeaderChecksum`, and this should not exist in the generic checksum code.

Reply:
Ah, I didn't realize the relationship there, or that TCP checksums of 0 are fine. I'll just revert all this now, in favor of a new PR. Thanks.